// press-ready/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.lint = exports.build = exports.inspectPDF = exports.isGhostscriptAvailable = exports.ghostScript = exports.isPdfFontsAvailable = exports.pdfInfo = exports.pdfFonts = void 0;
var pdffonts_1 = require("./pdffonts");
Object.defineProperty(exports, "pdfFonts", { enumerable: true, get: function () { return pdffonts_1.pdfFonts; } });
Object.defineProperty(exports, "pdfInfo", { enumerable: true, get: function () { return pdffonts_1.pdfInfo; } });
Object.defineProperty(exports, "isPdfFontsAvailable", { enumerable: true, get: function () { return pdffonts_1.isPdfFontsAvailable; } });
var ghostScript_1 = require("./ghostScript");
Object.defineProperty(exports, "ghostScript", { enumerable: true, get: function () { return ghostScript_1.ghostScript; } });
Object.defineProperty(exports, "isGhostscriptAvailable", { enumerable: true, get: function () { return ghostScript_1.isGhostscriptAvailable; } });
var inspectPDF_1 = require("./inspectPDF");
Object.defineProperty(exports, "inspectPDF", { enumerable: true, get: function () { return inspectPDF_1.inspectPDF; } });
var build_1 = require("./commands/build");
Object.defineProperty(exports, "build", { enumerable: true, get: function () { return build_1.build; } });
var lint_1 = require("./commands/lint");
Object.defineProperty(exports, "lint", { enumerable: true, get: function () { return lint_1.lint; } });
